core/src/graph/graph.rs

```diff
@@ -563,15 +615,82 @@ impl Graph {
 	/// Determine if page is full
 	///  aggressive:false -> use a simple heuristic based on the number of connections
 	///  aggressive:true  -> do actual compression to determine resulting actual page size
-	pub fn is_page_full(&self, page: &GraphPage, aggressive: bool) -> bool {
+	pub fn try_add_connection_to_page(
+		&self,
+		page: &mut GraphPage,
+		connection_id: &DsnpUserId,
+		aggressive: bool,
+		keys: &Option<ResolvedKeyPair>,
+		dsnp_version_config: Option<&DsnpVersionConfig>,
+	) -> Result<()> {
 		let connection_type = self.get_connection_type();
-		let max_connections_per_page = *PAGE_CAPACITIY_MAP.get(&connection_type).unwrap_or(&50);
+		let max_connections_per_page =
+			*PAGE_CAPACITY_MAP.get(&connection_type).unwrap_or_else(|| {
+				let mut capacities: Vec<&usize> = PAGE_CAPACITY_MAP.values().collect();
+				capacities.sort();
+				capacities.first().unwrap() // default: return smallest capacity value
+			});
 
 		if !aggressive {
-			return page.connections().len() >= max_connections_per_page
+			if page.connections().len() >= max_connections_per_page {
+				return Err(Error::msg("Page trivially full"))
+			}
+
+			if let Err(e) = page.add_connection(connection_id) {
+				return Err(e)
+			}
+
+			return Ok(())
+		}
+
+		if connection_type.privacy_type() == PrivacyType::Private &&
+			(dsnp_version_config.is_none() || keys.is_none())
+		{
+			return Err(Error::msg("Missing encryption key or config"))
+		}
+
+		let max_page_size = self.environment.get_config().max_graph_page_size_bytes as usize;
+		// Max page fullness leaves room for 2 uncompressed connections
+		let max_page_fullness = max_page_size -
+			(2 as usize *
+				(std::mem::size_of::<DsnpGraphEdge>() + std::mem::size_of::<DsnpPrid>()));
```

Comment:
There are cases that in memory size is different from serialized size due to padding and ... We would definitely need to check this with the actual serialized size.

Reply:
I'm not that concerned. Leaving space for 2 *uncompressed* connections is probably more than enough of a buffer, since (1) even worst case compression of a single connection won't double the size, and (2) we calculate page fullness using post-compression, post-encryption serialized data, so really the only reason to reserve space in the page would be to hedge against future on-chain modifications that introduce overhead to paginated storage.

I'm open to reducing the margin of error even further; but I think it's probably best to leave it as is for right now and revisit once we have real chain interactions to analyze.

Reply:
Removed extra space padding completely based on discussion.